text/0000-procedural-vtables.md

```diff
@@ -0,0 +1,490 @@
+work on the type. This data pointer adjustment is wide pointer specific and overridable via the `CustomProjectionToUnsized` trait.
+For regular fields the `CustomProjection` trait handles the extraction of the sized pointer to the field.
+
+## Traits managing the unsizing and projecting
+
+When unsizing, the `<dyn MyTrait as CustomUnsize>::unsize` method is invoked.
+The only reason that trait must be
+`impl const CustomUnsize` is to restrict what kind of things you can do in there, it's not
+strictly necessary. This restriction may be lifted in the future.
+
+For all other operations, the methods on `<dyn MyTrait as CustomUnsized>::WidePointer` are invoked.
+
+When obtaining function pointers from vtables, instead of computing an offset, the `MyTrait for dyn MyTrait` impl's
+methods are invoked, allowing users to insert their own logic for obtaining the runtime function pointer.
+Through the use of MIR optimizations (e.g. inlining), the final LLVM assembly is tuned to be exactly the same as today.
+The above statement contains significant hand-waving, but I propose we block the stabilization of this RFC on the
+relevant optimizations existing, which will then allow users to reproduce the performance of the built-in unsizing.
+
+These types' and trait's declarations are provided below:
+
+```rust
+unsafe trait CustomUnsized {
+    type WidePointer: Copy;
+    fn size_of(ptr: WidePointer) -> usize;
+    fn align_of(ptr: WidePointer) -> usize;
+}
+
+unsafe trait CustomUnsize<DynTrait> where DynTrait: CustomUnsized {
+    fn unsize<const owned: bool>(t: *mut Self) -> DynTrait::WidePointer;
+}
+
+unsafe trait CustomProjection: CustomUnsized {
+    /// The offset is in bytes.
+    unsafe fn project(ptr: <Self as CustomUnsized>::WidePointer, offset: usize) -> *mut u8;
+}
+
+unsafe trait CustomProjectionToUnsized: CustomUnsized {
+    /// The offset is in bytes and must be the exact offset from the start of the unsized struct to its unsized field.
+    unsafe fn project_unsized(ptr: <Self as CustomUnsized>::WidePointer, offset: usize) -> <Self as CustomUnsized>::WidePointer;
+}
+```
+
+The 
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+* This may be a serious case of overengineering. We're basically taking vtables out of the language and making dynamic dispatch on trait objects a user definable thing.
+* This may slow down compilation, likely entirely preventable by keeping a special case in the compiler for regular trait objects.
+* This completely locks us into never adding multiple vtable formats for a single trait. So you can't use a trait both as a C++ like vtable layout in some situations and a Rust wide pointer layout in others.
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+This is a frequently requested feature and as a side effect obsoletes `extern type`s which have all kinds of problems (like the inability to invoke `size_of_val` on pointers to them).
+
```

Comment:
Consider mentioning that we could instead let the language handle just the data pointer, ridding ourselves of the project and unsize functions at the cost of not controlling the entire pointer layout.

Reply:
In that case we'd go with any of the other RFCs ;) Which is something I do list further down.